### PR TITLE
Another batch of changes

### PR DESCRIPTION
**Pull request by tomers**
Each commit does something else, so please read each one separately

### CODE REVIEW

**Comment by maresb** [comment]
Thanks @tomers!

I'm not quite following what is `VERTICAL_PREVIEW_MARGIN_PX`. Could you explain it?

Is the idea that for the preview we add a fake margin for the nonprintable area so that the preview resembles the physical tape after printing?

**Comment by maresb** [comment]
That's really nice with the adjustable margin in the preview. But it's a bit weird that the margin code for the preview is different from the margin code for the printer.

Looking more closely I see that the printer code prints no margin at the beginning and a double-margin at the end. I think this probably has to do with the gap between the printer head and the cutter. I'm guessing that the gap is approximately `DEFAULT_MARGIN_PX` pixels long. Then when we print `2 * DEFAULT_MARGIN_PX`, where the first `DEFAULT_MARGIN_PX` is the true right-margin of the current label, and the second `DEFAULT_MARGIN_PX` is the left-margin of the next label. Do you agree with this assessment?

It might be nice to start distinguishing some of these concepts in the code. Like have a constant for `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX` = 13, and have separate variables in the code, something like `current_right_margin` and `future_left_margin` (or whatever makes sense), just to make the code somewhat consistent with reality. 

In principle we could enable a margin-free mode where you print the first `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX` pixels, tell the user to cut, and print the remaining pixels plus future margin, and tell the user to cut again.

If the user wants a label with `margin` greater than `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX`, then I think the correct way to do this is:
* Print a left margin of `additional_left_margin = margin - GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX`, since we already have `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX` due to the gap from the previous label.
* Print the main payload
* Print a right margin of `margin`
* Continue printing a margin of `future_left_margin = GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX`.
(I think the current implementation is buggy since it'll produce too short of a left-margin and too long of a right-margin, but I'm not able to test at the moment.)

Am I making sense?

**Comment by tomers** [comment]
@marseb, thanks for the insights. I've contrinued to modify the code, and I am going to introduce some further changes soon. Thanks for the insight of explaining why we print the margins twice at the end. I was puzzled by that.
I will add some of your text as comments in the code, and I also consider taking the variable and constant names you suggest. Let's keep this PR open for a while, until I'm done with that.

**Comment by maresb** [comment]
Awesome, thanks so much for looking into that!!!

I've converted the PR to draft for now. When you're finished just hit "Ready for review" and I'll give it another look and merge if I approve.

**Comment by tomers** [comment]
Exactly. I still need to work on that part, but the idea is that preview
should show both horizontal and vertical margins.

**Comment by maresb** [comment]
Okay, great!

Keep in mind that there are a few different levels of abstraction coming together:
* the bitmap itself that's to be printed
* the way the bitmap is printed onto a piece of tape

I think there are valid use cases to have preview functionality for both. Unfortunately I don't have any clever ideas for an implementation.

Note that this is starting to connect up with @tomek-szczesny's work on a lookup table for various printer and label width combinations in #91.

It's really great how things are coming together thanks to your work @tomers! Back when #91 was opened, I don't think the code was healthy enough to carry through that implementation. But now the code is becoming much easier to understand, opening up lots of exciting possibilities!

**Comment by tomek-szczesny** [comment]
Certainly I'm glad someone's cleaning up my amateur code. :) 
Just wanted to add, I regret that printer must be detected before rendering, in order to determine the actual printable area. I see no other way around it, barren the user-provided printer model as a command line parameter.

**Comment by maresb** [comment]
> my amateur code

To be clear the obstacle wasn't your code, it was instead the underlying code which was very idiosyncratic with not-so-great separation of concerns and idiosyncratic variable names. Those sorts of things build up and create a tremendous amount of friction for making changes.

**Comment by tomers** [comment]
Thanks, @maresb and @tomek-szczesny!
I'm working on discarding margin handling in the printing driver. This adds mental load on understanding the code, and the principle of having to end the print job with an offset for the next print job, and will be a burden in the future, when we'll have multiple drivers (which would need to duplicate this task), and also in the case where we may want to implement some more complicated scenarios, such as printing batch of labels, with zero margin (i.e. print brank without ending with offset, cut the label, then iteratively printing labels without margins, or at least with SW defined margins, and instructing user to cut).

@maresb, regarding your comment about idiosyncratic code, that's indeed true, and the was my initial effort, to clean these things off the table, before approaching any other new features.

File: src/dymoprint/lib/utils.py
Comment (maresb) [inline src/dymoprint/lib/utils.py:42]:
```suggestion
def mm_to_px(mm) -> int:
```

Comment (tomers) [reply src/dymoprint/lib/utils.py:42]: I prefer to have pixel as float still. It indeed makes no physical sense, but helps with calculations along the way. I prefer to convert to int at the very end.

Comment (tomers) [reply src/dymoprint/lib/utils.py:42]: Added one more commit. Please review.

**Comment by tomers** [comment]
Hey @maresb and @tomek-szczesny,

Can you please review this PR?

Thanks,
Tomer

File: src/dymoprint/lib/utils.py
Comment (maresb) [inline src/dymoprint/lib/utils.py]: If you want it to return a float:

```suggestion
    return mm * PIXELS_PER_MM
```

**Review by maresb — CHANGES_REQUESTED** [review]
I will finish a PR soon with some more in-depth suggestions

**Comment by maresb** [comment]
Oh, looks like you're actively working on this. I'll hold off on further review for the moment.

**Comment by maresb** [comment]
Are you okay with the conflicts there? Looks like maybe you accidentally rebased on a stale branch?

**Comment by tomers** [comment]
It is ready for review now. Sorry, but I use `git rebase` a lot, and I keep forgeting to rebase on `origin/master`.

**Comment by tomers** [comment]
Any news?

**Comment by maresb** [comment]
Sorry, I'm in a time crunch at the moment. It might have to wait another few days.

**Comment by tomers** [comment]
Hey there,

Due to irresponsivenes, I'm abandoning this project, and will fork it in a new organization, managed by me. Whoever wants to join me, is most welcome to contact me via LinkedIn.

**Comment by maresb** [comment]
Dang, that's pretty drastic. As you wish.

**Comment by tomers** [comment]
I hope I did not sound too drastic. That was not the intention. I mean that
I will cease development on that repository in the future.
I have no hard feelings at all! It is simply that I can't wait 3 weeks for
a PR to be merged... It's hard to work like that.
Note that my action is aligned to what we've discussed earlier this month
on LinkedIn. We agreed that we should open a separate, new, Github
organization for this purpose.

**Comment by maresb** [comment]
Ya, understandable. I'm going through a very stressful period at the moment, and unfortunately I haven't had any time in the past few weeks to commit to open-source.

I would have expected such an announcement to be a bit more coordinated and communal, not so unilateral. But you seem to have a lot of energy, so I think it's good that you to take charge.

**Comment by tomek-szczesny** [comment]
This repository is already abandoned by its actual owner so it does make sense to fork it into a new entity that will be easier to pass control to. I remember that @maresb had problems with privileges here already.

I think a link to the new repo / organization should be added on top of README here, and this repository archived if possible.

**Comment by tomers** [comment]
> I would have expected such an announcement to be a bit more coordinated
and communal, not so unilateral
I totally agree with that. I am sorry for being hasty with my response.
Ben and Tomek, maybe we should find some time for a short discussion (this
medium is bad for that) about how we're going to push this project forward.
Alternatively, if you approve, I can open a new Github organization and
fork this project (I will also update README accordingly).

**Comment by tomek-szczesny** [comment]
I don't feel there is anything left to discuss, a new organization and fork of this repository is reasonable and I think everyone has already expressed approval. Just let us know so we can join it. :)

**Comment by maresb** [comment]
I wanted to send an email to recent contributors, and also @computerlyrik, letting them know about the plan, and also to solicit a new name for the fork. There are also some logistical issues like how to communicate the deprecation of this project. I think a quick call may be useful to sort out these details. @tomek-szczesny, it'd be nice to have you there too, but @tomers and I could probably make do with just the two of us.

**Comment by tomers** [comment]
@tomek-szczesny, @maresb, let's continue this discussion in issue #114 .

**Comment by maresb** [comment]
Some preliminary review to fix some breakage that I'm seeing: https://github.com/tomers/dymoprint/pull/6

Here's the output of `dymoprint --preview asdf` on `v2.4.0`:

image

Here's the output on https://github.com/tomers/dymoprint/pull/6

image

I can't even take a screenshot of the full output. I understand that we're enlarging the preview to simulate the printable area, but I'm wondering what all this extra black space is about.

File: src/dymoprint/gui/gui.py
Comment (maresb) [inline src/dymoprint/gui/gui.py]: I read `print` as a verb, and so this reads like a method name rather than an attribute. Maybe rename to something like `label_bitmap_to_print`?

File: src/dymoprint/gui/q_dymo_labels_list.py
Comment (maresb) [inline src/dymoprint/gui/q_dymo_labels_list.py]: I'd strongly recommend against introducing this and the next blind except a few lines down. They make things extremely difficult to maintain down the line.

Even if it's a generic catch-all thing, at least make an effort to catch all the common errors and leave `BaseException` as a failsafe rather than a primary measure.

File: src/dymoprint/lib/dymo_labeler.py
Comment (maresb) [inline src/dymoprint/lib/dymo_labeler.py]: Unless there's something else forcing `tape_size_mm` to be a supported tape size, this should be a `ValueError` and not an `AssertionError`. Note that Python strips assert statements when run in optimized mode, so assert statements shouldn't be used for any sort of error handling.

(Note that given our installation instructions probably 100% of our users will not be running in optimized mode, so my statement is more on principle than in practice; semantically `ValueError` is more accurate than `AssertionError`.)

File: src/dymoprint/lib/unicode_blocks.py
Comment (maresb) [inline src/dymoprint/lib/unicode_blocks.py:18]: What are transparent pixels for? Probably some explanation of this and `LS` should be added to the code as comments.

File: src/dymoprint/lib/render_engines/print_preview.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/print_preview.py:70]: The result looks a bit off:
image

The text is white and difficult to read.
The vertical margin appears to be displayed horizontally under the vertical margin.

**Review by maresb — COMMENTED** [review]
To avoid merge conflicts, please first see my suggested changes under https://github.com/tomers/dymoprint/pull/6.

File: src/dymoprint/lib/render_engines/print_preview.py
Comment (tomers) [reply src/dymoprint/lib/render_engines/print_preview.py:70]: I fixed the coloring. I've hard-coded the colors for now. The reason I've use the original colors is because I worked in dark mode, and it looks different there. I tried to choose colors which would look good on both modes, but this is still a WIP.

> The vertical margin appears to be displayed horizontally under the vertical margin
I am not sure I understand this sentence. Can you elaborate?

**Comment by tomers** [comment]
Hello @maresb , thank you for reviewing my code! :-)
Apologies for the delay in addressing the issues raised. Your feedback guided me towards better solutions, and I appreciate your patience throughout the process.

**Comment by maresb** [comment]
More image mode headaches :dizzy_face: 

```python
$ dymoprint --browser asdf --verbose
[DEBUG] Read config file: /home/mares/.config/dymoprint.ini
[DEBUG] Demo mode: showing label..
[ERROR] not supported for this image mode
Traceback (most recent call last):
  File "/home/mares/repos/dymoprint/src/dymoprint/lib/utils.py", line 49, in system_run
    yield
  File "/home/mares/repos/dymoprint/src/dymoprint/cli/cli.py", line 343, in main
    run()
  File "/home/mares/repos/dymoprint/src/dymoprint/cli/cli.py", line 325, in run
    ImageOps.invert(bitmap).save(fp)
  File "/home/mares/micromamba/envs/dymoprint-dev/lib/python3.8/site-packages/PIL/ImageOps.py", line 563, in invert
    return image.point(lut) if image.mode == "1" else _lut(image, lut)
  File "/home/mares/micromamba/envs/dymoprint-dev/lib/python3.8/site-packages/PIL/ImageOps.py", line 60, in _lut
    raise OSError(msg)
OSError: not supported for this image mode
```

File: src/dymoprint/lib/render_engines/print_preview.py
Comment (maresb) [reply src/dymoprint/lib/render_engines/print_preview.py:70]: Ah, now I see. It is really difficult to find something that works in both.

I don't know much about PyQt5, but I wonder if there's some way to set the colors conditionally based on light vs dark.

File: src/dymoprint/gui/q_dymo_labels_list.py
Comment (maresb) [inline src/dymoprint/gui/q_dymo_labels_list.py]: Can we get by here with `RenderEngineException` or do we need `BaseException` to cover all the cases?

File: src/dymoprint/gui/q_dymo_labels_list.py
Comment (maresb) [inline src/dymoprint/gui/q_dymo_labels_list.py:177]: Can we get on the next line with `RenderEngineException` or do we need `BaseException` to cover all the cases?

**Comment by maresb** [comment]
The issue with the huge preview margins described in https://github.com/computerlyrik/dymoprint/pull/113#issuecomment-1975153786 is resolved. :rocket:

File: src/dymoprint/lib/render_engines/barcode_with_text.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/barcode_with_text.py:12]: This should fit on a single line. (Just remove the last trailing comma and reformat.)

File: src/dymoprint/lib/render_engines/margins.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/margins.py]: I'd recommend replacing the enum with a string literal for simplicity.

File: src/dymoprint/lib/render_engines/margins.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/margins.py]:
```suggestion
        justify: Literal["left", "center", "right"] = "center",
```

File: src/dymoprint/lib/render_engines/margins.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/margins.py]:
```suggestion
        # label's edge (the label's edge begins just under the cutter).
```

File: src/dymoprint/lib/render_engines/margins.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/margins.py]: This offset business is so confusing. I'm trying to make it a bit more clear:

```suggestion
        # Subsequently, in order to achieve symmetry so that the final margin matches
        # the initial margin, we add another offset of DX pixels, so that the cut
        # will have that same margin from both sides of the payload edge. In summary,
        # due to the gap between the printer head and the cutter of DX pixels,
        # printing an initial margin of 0 pixels and a final margin of 2 * DX pixels
        # leaves an effective margin of DX pixels on both sides.
```

File: src/dymoprint/lib/render_engines/print_payload.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/print_payload.py]: We should probably do a search through the codebase for "left" and "center" and make sure everything's typed correctly.
```suggestion
        justify: Literal["left", "center", "right"] = "center",
```

File: src/dymoprint/lib/render_engines/print_preview.py
Comment (maresb) [inline src/dymoprint/lib/render_engines/print_preview.py]:
```suggestion
        justify: Literal["left", "center", "right"] = "center",
```

File: src/dymoprint/lib/unicode_blocks.py
Comment (maresb) [inline src/dymoprint/lib/unicode_blocks.py]: I believe only the second pixel can be transparent, right?
```suggestion
    (WHITE_PIXEL, TRANSPARENT_PIXEL): NB,
```

We should probably be testing this with an odd-length payload that actually touches the edge.

File: src/dymoprint/lib/unicode_blocks.py
Comment (maresb) [inline src/dymoprint/lib/unicode_blocks.py]:
```suggestion
    (BLACK_PIXEL, TRANSPARENT_PIXEL): NB,
    (WHITE_PIXEL, TRANSPARENT_PIXEL): UH
```

File: src/dymoprint/lib/unicode_blocks.py
Comment (maresb) [inline src/dymoprint/lib/unicode_blocks.py]: Unfortunately it doesn't seem like there are unicode block characters that shade the upper and lower half independently.

It seems to me like we're probably better off leaving the transparent pixels unshaded.

Unless I'm confused!

```suggestion
```

**Review by maresb — CHANGES_REQUESTED** [review]

File: src/dymoprint/lib/render_engines/barcode_with_text.py
Comment (tomers) [reply src/dymoprint/lib/render_engines/barcode_with_text.py:12]: Ruff does not approve of your suggestion, and pre-commit fails.
Also, when I save in VS Code, the IDE auto formats the line for me.

File: src/dymoprint/lib/render_engines/margins.py
Comment (tomers) [reply src/dymoprint/lib/render_engines/margins.py]: I prefer not to. I think it is more obvious to have a single place where the possible values are.
I think this way it is better than having to add `Literal` type hint to each argument in the function call chain, where in each such type hint we have to specify all accepted values. This is what `Enum`s are for, no?

File: src/dymoprint/lib/unicode_blocks.py
Comment (tomers) [reply src/dymoprint/lib/unicode_blocks.py]: `dymoprint --preview-inverted "single line"` should cover it (in current setup it results in an off image)

File: src/dymoprint/lib/unicode_blocks.py
Comment (tomers) [reply src/dymoprint/lib/unicode_blocks.py]: I just added the shade for debugging purposes (to see where it is used), and probably forgot to remove it.

**Comment by tomers** [comment]
Done with the fixes, @maresb

File: src/dymoprint/lib/render_engines/margins.py
Comment (maresb) [reply src/dymoprint/lib/render_engines/margins.py]: Yes, that's what `Enum`s are for, but I disagree that it's more obvious to have the possible values in a single place.

Probably the most important metric of code quality to me is, "How much do I have to poke around in the code before I understand what's going on?" When using an `Enum` as opposed to `Literal` it means that the given block of code no longer stands on its own. Whenever I want to understand or use it (e.g. from a REPL) I must look into another file to read the definition and then either import the `Enum` or make use of the meaningless correspondence between names and values.

The obvious counterpoint is `Enum`s are good because they're DRY, and DRY is good. But why is DRY good? It usually reduces code length, but I prioritize clarity over length. It's also good to prevent different parts of code from going out-of-sync. But with `Literal` you have the type checker for that. And the type checker will do a much better job with `Literal` than `Enum`. If you add a new member to your `Enum` you should update all functions that use it. With `Literal` the type checker should automatically flag all the functions that need to be updated. It's also more powerful because for example it's easy to define functions which only handle subsets of the keys.

`Enum`s aren't terrible, but unless you have a large number of elements or if there's a meaningful correspondence between names and values, I just don't see any benefit. My feeling is that they're mostly obsolete in modern Python. Ultimately it's a subjective style choice, so feel free to pass if you're unconvinced.

File: src/dymoprint/lib/render_engines/barcode_with_text.py
Comment (maresb) [reply src/dymoprint/lib/render_engines/barcode_with_text.py:12]: Indeed. Somehow the line looked deceivingly short to me.

File: src/dymoprint/lib/render_engines/margins.py
Comment (maresb) [reply src/dymoprint/lib/render_engines/margins.py]: What do you think of this?

File: src/dymoprint/lib/render_engines/margins.py
Comment (maresb) [reply src/dymoprint/lib/render_engines/margins.py]: And this?

File: src/dymoprint/lib/render_engines/print_payload.py
Comment (maresb) [reply src/dymoprint/lib/render_engines/print_payload.py]: There's still one missing in `dymoprint.gui.q_dymo_labels_list` line 116.

**Comment by maresb** [comment]
`--browser` now works again. There are still a few things unresolved from the last review.

It's really difficult to review such a large PR after a rebase, so could you please refrain from rebasing until after I've approved?

File: src/dymoprint/lib/render_engines/print_preview.py
Comment (tomers) [reply src/dymoprint/lib/render_engines/print_preview.py:70]: I added a commit which addresses this issue

File: src/dymoprint/lib/render_engines/print_payload.py
Comment (tomers) [reply src/dymoprint/lib/render_engines/print_payload.py]: Fixed.
That's exactly what I dislike about working wet, without Enums... You have to update 100 places.
I've searched in the code, and I think I've updated in all places.

File: src/dymoprint/lib/render_engines/margins.py
Comment (tomers) [reply src/dymoprint/lib/render_engines/margins.py]: Perfect. Thanks.

File: src/dymoprint/lib/render_engines/margins.py
Comment (tomers) [reply src/dymoprint/lib/render_engines/margins.py]: I've commited a change to use literal type hint instead of enum. I feel less confident relying on the liner for detecting issues such as we had earlier during review, when in some places I've forgot to update the type literal. I personally favor DRY code (I really struggle not to fix wet code whenever I encounter it), which I believe improves code's maintainability. I think that prioritizing maintanability over readability (simplicity of the code) is the way to go. However, in this project, I do not want to enforce my personal preference, and I have no problem compromising to get to a consensus, which I believe would increase productivity and improve overall developer experience of all of us :-)

File: src/dymoprint/lib/render_engines/print_payload.py
Comment (maresb) [reply src/dymoprint/lib/render_engines/print_payload.py]: Seems to me like a feature. Sometimes a change does require an update in 100 places, and this makes it explicit. :slightly_smiling_face:

**Review by maresb — APPROVED** [review]
Whew, that was a whole lot of work, but I think we finally got it. Thanks @tomers for persevering. These are some really huge improvements.

Go ahead and rebase if you'd like, and then let's get this merged.

**Comment by tomers** [comment]
Thanks Ben,

The branch is ready to be merged from my side. I am not going to change it
anymore. You can merge according to the plan that we discussed privately.

**Comment by maresb** [comment]
Thanks @tomers!

For anyone reading, to avoid giving the impression of being secretive, the plan is to not release this as DymoPrint, but instead release it as Labelle v1.1.0 (see #114). And Labelle v1.0.0 will be a rebranded DymoPrint v2.4.0.

**Pull request merged**
